src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCheckedParameterTests.cs

```diff
@@ -0,0 +1,456 @@
+  IL_0011:  throw
+ -IL_0012:  nop
+ -IL_0013:  ret
+  }", sequencePoints: "C.M");
+        }
+
+        [Fact]
+        public void TestNullCheckedIndexedProperty()
+        {
+            var source = @"
+class C
+{
+    public string this[int index!] => null;
+    public static void Main() { }
+}";
+
+            // Release
+            var compilation = CompileAndVerify(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyIL("C.this[int].get", @"
+{
+  // Code size       11 (0xb)
+  .maxstack  1
+ ~IL_0000:  ldarg.1
+  IL_0001:  brtrue.s   IL_0009
+  IL_0003:  newobj     ""System.Exception..ctor()""
+  IL_0008:  throw
+ -IL_0009:  ldnull
+  IL_000a:  ret
+}", sequencePoints: "C.get_Item");
+
+            // Debug
+            compilation = CompileAndVerify(source, options: TestOptions.DebugExe);
+            compilation.VerifyIL("C.this[int].get", @"
+{
+   // Code size       11 (0xb)
+   .maxstack  1
+  ~IL_0000:  ldarg.1
+   IL_0001:  brtrue.s   IL_0009
+   IL_0003:  newobj     ""System.Exception..ctor()""
+   IL_0008:  throw
+  -IL_0009:  ldnull
+   IL_000a:  ret
+}", sequencePoints: "C.get_Item");
+        }
+
+        [Fact]
+        public void TestNullCheckedIndexedGetterSetter()
+        {
+            var source = @"
+class C
+{
+    private string[] names;
+    public string this[int index!]
+    {
+        get
+        {
+            return names[index];
+        }
+        set
+        {
+            names[index] = value;
+        }
+    }
+    public static void Main() { }
+}";
+            // Release
+            var compilation = CompileAndVerify(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyIL("C.this[int].set", @"
+{
+    // Code size       19 (0x13)
+    .maxstack  3
+    ~IL_0000:  ldarg.1
+    IL_0001:  brtrue.s   IL_0009
+    IL_0003:  newobj     ""System.Exception..ctor()""
+    IL_0008:  throw
+    -IL_0009:  ldarg.0
+    IL_000a:  ldfld      ""string[] C.names""
+    IL_000f:  ldarg.1
+    IL_0010:  ldarg.2
+    IL_0011:  stelem.ref
+    -IL_0012:  ret
+}", sequencePoints: "C.set_Item");
+
+            // Debug
+            compilation = CompileAndVerify(source, options: TestOptions.DebugExe);
+            compilation.VerifyIL("C.this[int].set", @"
+{
+    // Code size       20 (0x14)
+    .maxstack  3
+    ~IL_0000:  ldarg.1
+    IL_0001:  brtrue.s   IL_0009
+    IL_0003:  newobj     ""System.Exception..ctor()""
+    IL_0008:  throw
+    -IL_0009:  nop
+    -IL_000a:  ldarg.0
+    IL_000b:  ldfld      ""string[] C.names""
+    IL_0010:  ldarg.1
+    IL_0011:  ldarg.2
+    IL_0012:  stelem.ref
+    -IL_0013:  ret
+}", sequencePoints: "C.set_Item");
+        }
+
+        [Fact]
+        public void TestNullCheckedIndexedSetter()
+        {
+            var source = @"
+class C
+{
+    private string[] names;
+    public string this[int index!]
+    {
+        set
+        {
+            names[index] = value;
+        }
+    }
+    public static void Main() { }
+}";
```

Comment:
Consider simplifying this test to:
```C#
class C
{
    object this[object index!] { set { } }
}
```
It's shorter and it would also test the case where the method body is empty.